cumulusci/core/github.py

```diff
@@ -268,3 +303,131 @@ def get_version_id_from_tag(repo: Repository, tag_name: str) -> str:
+    @returns: The formatted exception string
+    """
+    user_warning = ""
+    unauthorized_warning = """
+    Bad credentials. Verify that your personal access token is correct and that
+    you are authorized to access this resource.
+    """
+
+    too_many_str = "too many 401 error responses"
+    is_bad_auth_retry = (
+        type(exc) is TransportError
+        and type(exc.exception) is RetryError
+        and too_many_str in str(exc.exception)
+    )
+    is_auth_failure = type(exc) is AuthenticationFailed
+
+    if is_bad_auth_retry or is_auth_failure:
+        user_warning = unauthorized_warning
+    else:
+        scope_error_msg = check_github_scopes(exc)
+        sso_error_msg = check_github_sso_auth(exc)
+        user_warning = scope_error_msg + sso_error_msg
+
+    return user_warning
+
+
+def check_github_scopes(exc: ResponseError) -> str:
+    """
+    Parse github3 ResponseError headers for the correct scopes and return a
+    warning if the user is missing.
+
+    @param exc: The exception to process
+    @returns: The formatted exception string
+    """
+
+    user_warning = ""
+
+    has_wrong_status_code = exc.response.status_code not in (403, 404)
+    if has_wrong_status_code:
+        return user_warning
+
+    token_scopes = get_oauth_scopes(exc.response)
+
+    # Gist resource won't return X-Accepted-OAuth-Scopes for some reason, so this
+    # string might be `None`; we discard the empty string if so.
+    accepted_scopes = exc.response.headers.get("X-Accepted-OAuth-Scopes") or ""
+    accepted_scopes = set(accepted_scopes.split(", "))
+    accepted_scopes.discard("")
+
+    request_url = urlparse(exc.response.url)
+    if not accepted_scopes and request_url.path == "/gists":
+        accepted_scopes = {"gist"}
+
+    missing_scopes = accepted_scopes.difference(token_scopes)
+    if missing_scopes:
+        user_warning = f"Your token may be missing the following scopes: {', '.join(missing_scopes)}\n"
+        # This assumes we're not on enterprise and 'api.github.com' == request_url.hostname
+        user_warning += "Visit https://github.com/settings/tokens to add them."
+
+    return user_warning
+
+
+def check_github_sso_auth(exc: ResponseError) -> str:
+    """
+    Check ResponseError header for SSO authorization and return a warning if
+    required
+
+    @param exc: The exception to process
+    @returns: The formatted exception string
+    """
+    user_warning = ""
+    headers = exc.response.headers
+
+    if exc.response.status_code != 403 or "X-Github-Sso" not in headers:
+        return user_warning
+
+    sso_header = str(headers["X-Github-Sso"] or "")
+    if sso_header.startswith("required; url="):
+        # In this case the message from github is good enough, but we can help
+        # the user by opening a browser to authorize the token.
+        auth_url = sso_header.split("url=", maxsplit=1)[1]
+        user_warning = f"{exc.message}\n{auth_url}"
+        webbrowser.open(auth_url)
```

Comment:
What happens if we're running headless?

Reply:
tl;dr: returns `False`

---
Google fail, had to do it the hard way.

Image is `python:3.8-slim-buster`:

```console
❯ docker run -it empty-python
Python 3.8.11 (default, Jul 22 2021, 15:41:12)
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import webbrowser
>>> webbrowser.open('google.com')
False
>>>
❯ python
Python 3.9.5 (default, Jun  7 2021, 17:45:25)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import webbrowser
>>> webbrowser.open('google.com')
True
```